### PR TITLE
Fix issue with lemma during equality engine iterator in sets

### LINKED ISSUE #4370: Fatal failures at src/theory/uf/equality_engine_iterator.cpp:104
Hi,
For the following formula:
~~~~
(declare-const v6 Bool)
(declare-const i4 Int)
(declare-const i7 Int)
(declare-const i10 Int)
(declare-const i11 Int)
(declare-fun st0 () (Set Int))
(declare-fun st1 () (Set Int))
(declare-fun st3 () (Set Int))
(declare-fun st4 () (Set Int))
(declare-fun st5 () (Set Int))
(declare-fun st7 () (Set Int))
(declare-fun st8 () (Set Int))
(declare-fun st12 () (Set Int))
(declare-const i12 Int)
(declare-const v19 Bool)
(declare-const i16 Int)
(declare-const i17 Int)
(declare-const v71 Bool)
(assert (> (card (intersection (union st7 st1) st1)) 0))
(assert (! (or (subset st1 st7) (subset st7 st4) (subset st3 st5)) :named IP_216))
(check-sat-assuming ((! (or v71 (= (- (abs i11)) 402) (=> v6 (<= 731 396))) :named IP_112) (! (or (or (subset (union (union st7 st1) st1) st7) (> (- i4 33 i4 0) 0) (= (xor true (distinct 24 (abs 449)) true true (> 85 396)) (<= 469 33) true false (< 731 396)) (< (* i17 i16 i4 (* (abs i11) (- i4 33 i4 0) i12 i7 i12)) 182) (subset st7 (intersection (intersection st4 st0) st12)) (subset (union (union st3 st8) st8) st5) (> 85 396) false (=> true false) false false) v19 (> 986 i10)) :named IP_116)))
~~~~

CVC4 throws out a fatal failure at equality_engine.cpp:2387 (--sets-ext)
~~~~
Fatal failure within CVC4::theory::eq::EqClassIterator& CVC4::theory::eq::EqClassIterator::operator++() at /home/peisen/test/tofuzz/CVC4/src/theory/uf/equality_engine.cpp:2387
Check failure

 d_start == d_ee->getEqualityNode(d_current).getFind()

Aborted (core dumped)

~~~~

OS: Ubuntu 16.04
Commit: ad7907adf

>

### PR DESCRIPTION
**Pull request by ajreynol**
Fixes #4370.

### CODE REVIEW

File: src/theory/sets/theory_sets_private.cpp [hunk 1]
```diff
@@ -286,6 +286,10 @@ void TheorySetsPrivate::fullEffortCheck()
           d_card_enabled = true;
           // register it with the cardinality solver
           d_cardSolver->registerTerm(n);
+          if (d_im.hasSent())
+          {
+            break;
```
Comment (mudathirmahgoub) [inline src/theory/sets/theory_sets_private.cpp:291]: Why does this break solve the issue,  when there is a  new lemma  or fact?

Comment (ajreynol) [reply src/theory/sets/theory_sets_private.cpp:291]: `d_cardSolver->registerTerm` may send a lemma, which then calls Theory::preRegisterTerm, which in turn modifies the equality engine, making the equality engine iterator we are using invalid.

Comment (mudathirmahgoub) [reply src/theory/sets/theory_sets_private.cpp:291]: Thanks for the clarification.

**Review by mudathirmahgoub — APPROVED** [review]
LGTM

**Review by mudathirmahgoub — APPROVED** [review]

**Pull request merged**